### PR TITLE
make `memorynew` intrinsic

### PR DESCRIPTION
**Pull request by oscardssmith**
Attempt to split up https://github.com/JuliaLang/julia/pull/55913 into 2 pieces. This piece now only adds the `memorynew` intrinsic without any of the optimizations enabled by https://github.com/JuliaLang/julia/pull/55913. As such, this PR should be ready to merge now. (and will make https://github.com/JuliaLang/julia/pull/55913 smaller and simpler)
```
julia> f(T) = Memory{T}(undef, 2);

julia> @btime f(Int);
  7.879 ns (1 allocation: 48 bytes) #before
  3.959 ns (1 allocation: 48 bytes) #after 

julia> @btime f(Union{Int, Nothing})
  8.426 ns (1 allocation: 48 bytes) #before
  4.017 ns (1 allocation: 48 bytes) #after 

julia> @btime f(Any);
  8.637 ns (1 allocation: 48 bytes) #before
  3.920 ns (1 allocation: 48 bytes) #after 
```

### CODE REVIEW

File: src/common_symbols2.inc [hunk 1]
```diff
@@ -246,3 +246,5 @@ jl_symbol("invokelatest"),
 jl_symbol("jl_array_del_end"),
 jl_symbol("_mod64"),
 jl_symbol("parameters"),
+jl_symbol("monotonic"),
+jl_symbol("regex.jl"),
```
Comment (KristofferC) [inline src/common_symbols2.inc:250]: related to the `memorynew` intrinsic?

File: src/pipeline.cpp [hunk 2]
```diff
@@ -527,6 +527,7 @@ static void buildIntrinsicLoweringPipeline(ModulePassManager &MPM, PassBuilder *
             JULIA_PASS(FPM.addPass(LateLowerGCPass()));
             JULIA_PASS(FPM.addPass(FinalLowerGCPass()));
             if (O.getSpeedupLevel() >= 2) {
+                FPM.addPass(DSEPass());
```
Comment (KristofferC) [inline src/pipeline.cpp:530]: If this PR is only about adding the intrinsic itself, should it also add new optimization passes?

File: src/rtutils.c [hunk 3]
```diff
@@ -157,6 +157,13 @@ JL_DLLEXPORT void JL_NORETURN jl_has_no_field_error(jl_datatype_t *t, jl_sym_t *
     jl_throw(jl_new_struct(jl_fielderror_type, t, var));
 }
 
+JL_DLLEXPORT void JL_NORETURN jl_argument_error(char *str) // == jl_exceptionf(jl_argumenterror_type, "%s", str)
```
Comment (KristofferC) [inline src/rtutils.c:160]: What calls this?

File: test/core.jl [hunk 4]
```diff
@@ -4505,6 +4505,15 @@ for T in (Any, ValueWrapper)
     end
 end
 
+#test grow_end ccall directly since it's used in the C source
+for ET in [Nothing, Int, Union{Int, Nothing}, Any]
```
Comment (KristofferC) [inline test/core.jl:4509]: Is this test related to the new intrinsic?

File: base/genericmemory.jl [hunk 5]
```diff
@@ -241,9 +241,10 @@ getindex(A::Memory, c::Colon) = copy(A)
 
 ## Indexing: setindex! ##
 
-function _setindex!(A::Memory{T}, x::T, i1::Int) where {T}
-    ref = memoryrefnew(memoryref(A), i1, @_boundscheck)
-    memoryrefset!(ref, x, :not_atomic, @_boundscheck)
+function _setindex!(A::Memory{T}, x::T, i::Int) where {T}
+    @boundscheck Core.Intrinsics.ult_int(i, A.length)
```
Comment (KristofferC) [inline base/genericmemory.jl:245]: Why does this add a bounds check now that wasn't there before?

File: base/essentials.jl [hunk 6]
```diff
@@ -382,8 +382,17 @@ default_access_order(a::GenericMemory{:atomic}) = :monotonic
 default_access_order(a::GenericMemoryRef{:not_atomic}) = :not_atomic
 default_access_order(a::GenericMemoryRef{:atomic}) = :monotonic
 
-getindex(A::GenericMemory, i::Int) = (@_noub_if_noinbounds_meta;
-    memoryrefget(memoryrefnew(memoryrefnew(A), i, @_boundscheck), default_access_order(A), false))
+# bootstrap version for Memory{Any}
```
Comment (KristofferC) [inline base/essentials.jl:385]: I don't really understand this comment. What is the commented code used for? It is identical to what used to be there except with `Memory{Any}`?

File: doc/src/manual/performance-tips.md [hunk 7]
```diff
@@ -1058,12 +1058,12 @@ the output. As a trivial example, compare
 
 ```jldoctest prealloc
 julia> function xinc(x)
-           return [x, x+1, x+2]
+           return [x + i for i  in 1:3000]
```
Comment (KristofferC) [inline doc/src/manual/performance-tips.md:1061]:
```suggestion
           return [x + i for i in 1:3000]
```

File: doc/src/manual/performance-tips.md [hunk 8]
```diff
@@ -1075,16 +1075,16 @@ with
 
 ```jldoctest prealloc
 julia> function xinc!(ret::AbstractVector{T}, x::T) where T
-           ret[1] = x
```
Comment (KristofferC) [inline doc/src/manual/performance-tips.md:1078]: When there wasn't a loop it felt ok to "hardcode" the length to 3 but now it seems better to use the length of `ret`?

File: test/core.jl (hunk 4 above)
Comment (oscardssmith) [reply test/core.jl:4509]: yes (in that the refactoring in genericmemory.c initially broke this test)

File: src/rtutils.c (hunk 3 above)
Comment (oscardssmith) [reply src/rtutils.c]: oops. was supposed to be dropped from this.

File: src/pipeline.cpp (hunk 2 above)
Comment (oscardssmith) [reply src/pipeline.cpp]: agreed.

File: src/common_symbols2.inc (hunk 1 above)
Comment (oscardssmith) [reply src/common_symbols2.inc]: no.

File: base/essentials.jl (hunk 6 above)
Comment (oscardssmith) [reply base/essentials.jl]: yeah that was a mistake.

File: doc/src/manual/performance-tips.md (hunk 8 above)
Comment (oscardssmith) [reply doc/src/manual/performance-tips.md:1078]: agreed

File: base/genericmemory.jl (hunk 5 above)
Comment (Seelengrab) [reply base/genericmemory.jl:245]: Does this also need  `@propagate_inbounds`, or what the equivalent at that point of compilation is?

**Comment by gbaraldi** [comment]
This is much worse than the current implementation btw. For this you at least need a specialized builtin implementation in codegen, even if it just forwards the arguments.

**Comment by oscardssmith** [comment]
To prevent this PR from being a regression (and to fix the LLVM names test, I think the right way to go is to add the dynamic length version of codegen to this PR. It always goes through C (so LLVM won't be able to delete the whole allocation), but this way this PR on it's own is ~3ns faster to allocate arrays than master without the glory/risk of LLVM deleting the allocation entirely.

**Comment by oscardssmith** [comment]
@KristofferC I don't think this PR needs a pkgeval. It doesn't have any of the risks of https://github.com/JuliaLang/julia/pull/55913 wrt weird miscompiles.

**Comment by KristofferC** [comment]
I am surprised you would say that and I strongly disagree. The other PR was also just about to be merged before PkgEval showed it was quite buggy so it would be unfortunate to repeat the same mistake again.

**Comment by KristofferC** [comment]
> This is much worse than the current implementation btw. For this you at least need a specialized builtin implementation in codegen, even if it just forwards the arguments.

Has this been done now? It is hard to follow the evolution of the PR when the original commit gets amended over and over.

**Comment by KristofferC** [comment]
Also, benchmarks should be run here since it touches some core array functionality and to show the performance benefit and that no regressions gets introduced (e.g. from the seemingly new boundscheck).

File: Compiler/src/tfuncs.jl [hunk 9]
```diff
@@ -2244,7 +2250,16 @@ function _builtin_nothrow(𝕃::AbstractLattice, @nospecialize(f::Builtin), argt
...
-    if f === memoryrefnew
+    if f === Core.memorynew
+        argtypes[1] isa Const && argtypes[2] isa Const || return false
+        MemT = argtypes[1].val
+        isconcretetype(MemT) && MemT <: GenericMemory || return false
+        len = argtypes[2].val
+        len isa Int && 0 <= len < typemax(Int) || return false
+        elsz = datatype_layoutsize(MemT)
+        checked_smul_int(len, elsz)[2] && return false
+        return true
```
Comment (vchuravy) [inline Compiler/src/tfuncs.jl:2261]: I find this kind of "compressed" code quite difficult to read and reason about. 
The rest of this function outlines the more complicated checks into their own functions?

File: src/rtutils.c (hunk 3 above)
Comment (vchuravy) [reply src/rtutils.c:160]: Still here ;)

File: src/codegen.cpp [hunk 10]
```diff
@@ -1574,6 +1603,7 @@ static const auto &builtin_func_map() {
           { jl_f_nfields_addr,            new JuliaFunction<>{XSTR(jl_f_nfields), get_func_sig, get_func_attrs} },
           { jl_f__expr_addr,              new JuliaFunction<>{XSTR(jl_f__expr), get_func_sig, get_func_attrs} },
           { jl_f__typevar_addr,           new JuliaFunction<>{XSTR(jl_f__typevar), get_func_sig, get_func_attrs} },
+          { jl_f_memorynew_addr, new JuliaFunction<>{XSTR(jl_f_memorynew), get_func_sig, get_func_attrs} },
```
Comment (vchuravy) [inline src/codegen.cpp:1606]: Formatting seems off

**Comment by oscardssmith** [comment]
> The whole emit_memorynew looks like it could cause miscompilation issues?

@KristofferC the reason I say this version is much lower risk is that none of the segfaults that we were seeing come from the `emit_memorynew` function. (i.e. if you deleted the LLVM optimization passes we do, it would compile correctly). The cause of potential miscompiles in the other PR comes from the fact that if we teach the compiler how to do a Memory allocation completely (i.e. without the ccall that exists in this version), the compiler is able to analyze it and remove the allocation if it believes the allocation doesn't leak. This version of the PR keeps it as an "opaque" object allocation so the class of miscompile we were seeing can't occur. (and hence I think the regular Julia tests are likely to be sufficient for testing the PR).

> Has this been done now

yes.

File: src/codegen.cpp [hunk 11]
```diff
@@ -4421,6 +4451,20 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
...
+    else if (f == jl_builtin_memorynew && (nargs == 2)) {
+        const jl_cgval_t &memty = argv[1];
+        if (!memty.constant)
+            return false;
+        jl_datatype_t *typ = (jl_datatype_t*) memty.constant;
+        if (!jl_is_concrete_type((jl_value_t*)typ) || !jl_is_genericmemory_type(typ))
+            return false;
+        jl_genericmemory_t *inst = (jl_genericmemory_t*)((jl_datatype_t*)typ)->instance;
+        if (inst == NULL)
+            return false;
```
Comment (topolarity) [inline src/codegen.cpp:4463]: When does this happen?

Comment (oscardssmith) [reply src/codegen.cpp:4463]: this means the type isn't `Const` (i.e. for code like
`Memory{rand((Int, Float64))}(undef, 1)`

File: src/cgutils.cpp [hunk 12]
```diff
@@ -4427,6 +4432,107 @@ static int compare_cgparams(const jl_cgparams_t *a, const jl_cgparams_t *b)
...
+    Value *prod_with_overflow = ctx.builder.CreateCall(intr, {nel_unboxed, cg_elsz});
+    Value *nbytes = ctx.builder.CreateExtractValue(prod_with_overflow, 0);
+    Value *overflow = ctx.builder.CreateExtractValue(prod_with_overflow, 1);
+    if (isunion) {
+        // if isunion, we need to allocate the union selector bytes as well
+        intr = Intrinsic::getDeclaration(jl_Module, Intrinsic::sadd_with_overflow, ArrayRef<Type*>(T_size));
+        Value *add_with_overflow = ctx.builder.CreateCall(intr, {nel_unboxed, nbytes});
+        nbytes = ctx.builder.CreateExtractValue(add_with_overflow, 0);
+        Value *overflow1 = ctx.builder.CreateExtractValue(add_with_overflow, 1);
+        overflow = ctx.builder.CreateOr(overflow, overflow1);
```
[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/cgutils.cpp:4495]: sure (although just being able to build Julia is a pretty good test that none of these cases are broken)

File: src/codegen.cpp (hunk 11 above)
Comment (topolarity) [reply src/codegen.cpp:4463]: Isn't that the check 3 lines above?

Comment (oscardssmith) [reply src/codegen.cpp:4463]: oh, right. this case might be for `Memory{1}(undef, 1)` or something like that where it is Const, but not valid.

**Comment by oscardssmith** [comment]
@nanosoldier `runtests()`

**Comment by oscardssmith** [comment]
turns out that we want https://github.com/JuliaLang/julia/pull/56817 before this PR anyway since the segfaults on the OG memorynew PR were just a bug that cleaning up the IR magically is enough to trigger.

**Comment by vchuravy** [comment]
> and hence I think the regular Julia tests are likely to be sufficient for testing the PR

You are still adding more precise side effects in the modelling of the intrinsic that could have knock on effects down the road. I agree that this PR is lower risk, but PkgEval is not that expensive and it really helps being able to have that strong signal.

**Comment by maleadt** [comment]
> turns out that we want #56817 before this PR anyway since the segfaults on the OG memorynew PR were just a bug that cleaning up the IR magically is enough to trigger.

Better wait before triggering PkgEval then? I had to restart the server, so your submission was lost; you'll need to re-trigger (either right away, or after #56817 is merged).

**Comment by oscardssmith** [comment]
@nanosoldier `runtests()`

File: src/cgutils.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/cgutils.cpp:4495]: I've actually realized that this is fairly difficult to test for directly in ways that the compiler won't cheat at. i.e. inference knows that `memorynew(Memory{Int}, 5)` returns a `Memory{Int}`, so testing return type will just be optimized out unless I go through a bunch of effort to try to fool the compiler (which also definitely would have segfaulted if it couldn't allocate a `Memory`).

**Comment by oscardssmith** [comment]
no clue what's up with packageeval. if it doesn't come back within the next day, I'm merging without it. I have at least 1 PR (and hopefully another few after) that depend on this that I'm hoping to land in the 1.12 window.

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56803#issuecomment-2541619313) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/d77bb2a_vs_ece1c70/report.html) is available.

**Comment by oscardssmith** [comment]
failures look unreladed.

**Comment by gbaraldi** [comment]
LGTM

**Comment by KristofferC** [comment]
How was it determined that all the new segfault (that seem to happen in the same way) are unrelated to this?

Also, the performance benchmarks that were requested were not run.

What is even going on here?

**Comment by KristofferC** [comment]
> @KristofferC the reason I say this version is much lower risk... 

The previous PR was also about to be merged without any PkgEval so presumably the same risk assessment was made there.

**Comment by oscardssmith** [comment]
> How was it determined that all the new segfault (that seem to happen in the same way) are unrelated to this?

see discussion in https://julialang.slack.com/archives/C02C4RYM66N/p1734377710279309. The problem was stdlib migration breaking tarball signatures. The only differences between this PR when it was fully green on CI and when it merged to master was merge commits incorporating changes from master.

> Also, the performance benchmarks that were requested were not run.

Sorry about this. The performance impact is 3ns faster per Memory creation with no other changes I'm aware of.
> The previous PR was also about to be merged without any PkgEval

It turns out that the Pkgeval on that actually only found problems unrelated to that PR (see https://github.com/JuliaLang/julia/pull/56817 for the fix to the discovered issue)

**Comment by KristofferC** [comment]
> The performance impact is 3ns faster per Memory creation with no other changes I'm aware of.

Well yeah, presumably because that is the one benchmark that has been run. The point of running the full benchmark suite is to get aware of any other changes as well.

**Comment by KristofferC** [comment]
> . The problem was stdlib migration breaking tarball signatures 

How does that cause a bunch of new segfaults, e.g

```
59] signal 11 (128): Segmentation fault
in expression starting at /home/pkgeval/.julia/packages/Isoplot/g9Zby/test/runtests.jl:363
promote at ./promotion.jl:401
unknown function (ip: 0x7f91c57f2413) at (unknown file)
_jl_invoke at /source/src/gf.c:3358 [inlined]
ijl_apply_generic at /source/src/gf.c:3546
jl_apply at /source/src/julia.h:2235 [inlined]
jl_f__call_in_world_total at /source/src/builtins.c:921
_jl_invoke at /source/src/gf.c:3358 [inlined]
ijl_apply_generic at /source/src/gf.c:3546
jl_apply at /source/src/julia.h:2235 [inlined]
```

?

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs="@13f446e9f4d518b20faea68d2e239fcabd462ea9")`

**Comment by KristofferC** [comment]
I see that all the segfaulting packages have LoopVectorization or VectorizationBase as dependencies so hopefully, the crashes have something to do with that (in which case they can be ignored).

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56803#issuecomment-2548107743) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/da5acf6_vs_13f446e/report.md).

**Comment by oscardssmith** [comment]
looks like a bunch of improvements, a lot of noise on scalar benchmarks and a possible mild regression on Union arrays, (but it's somewhat unclear whether those were noise or not).

**Comment by vtjnash** [comment]
> @nanosoldier `runbenchmarks(ALL, vs="@13f446e9f4d518b20faea68d2e239fcabd462ea9")`

Very impressive

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56803#issuecomment-2550200495) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/da5acf6_vs_13f446e/report.md).

**Comment by vtjnash** [comment]
Seems like admits that sea of green, these 2 are red?

Test | TIME | MEM
-- | -- | --
["union", "array", ("broadcast", "*", "Int64", "(false, false)")] | 1.53 (5%) ❌ | 1.00 (1%)
["union", "array", ("broadcast", "*", "Int8", "(false, false)")] | 1.56 (5%) ❌ | 1.00 (1%)
["union", "array", ("broadcast", "*", "Int64", "(false, false)")] | 1.53 (5%) ❌ | 1.00 (1%)
["union", "array", ("broadcast", "*", "Int8", "(false, false)")] | 1.68 (5%) ❌ | 1.00 (1%)

**Comment by oscardssmith** [comment]
agreed. LLVM probably doesn't like the way that we're computing the size for the union case as much (or it's that inlining changes happened to get unlucky in 4 out of 100 cases)
